src/ape/managers/project.py

```diff
@@ -4,28 +4,257 @@
+    ``ApeProject`` at the current-working directory.
+    """
+
+    @property
+    def is_valid(self) -> bool:
+        if (self.path / CONFIG_FILE_NAME).exists():
+            return True
+
+        logger.warning(f"'{self.path.name}' is not an 'ApeProject'.")
+
+        # NOTE: We always return True as a last-chance attempt because it often
+        # works anyway and prevents unnecessary plugin requirements.
+        return True
+
+    @property
+    def sources(self) -> List[Path]:
+        """
+        All the source files in the project.
+        Excludes files with extensions that don't have a registered compiler.
+
+        Returns:
+            List[pathlib.Path]: A list of a source file paths in the project.
+        """
+
+        files: List[Path] = []
+        if not self.contracts_folder.exists():
+            return files
+
+        for extension in self.compilers.registered_compilers:
+            files.extend(self.contracts_folder.rglob(f"*{extension}"))
+
+        return files
+
+    def create_manifest(
+        self, file_paths: Optional[List[Path]] = None, use_cache: bool = True
+    ) -> PackageManifest:
+        # Create a config file if one doesn't exist to forward values from
+        # the root project's 'ape-config.yaml' 'dependencies:' config.
+        config_file = self.path / CONFIG_FILE_NAME
+        if not config_file.exists():
+            config_data = {}
+            if self.name:
+                config_data["name"] = self.name
+            if self.version:
+                config_data["version"] = self.version
+            if self.contracts_folder != "contracts":
+                # Only sets when not default.
+                config_data["contracts_folder"] = self.contracts_folder.name
+            if config_data:
+                with open(config_file, "w") as f:
+                    yaml.safe_dump(config_data, f)
+
+        # Load a cached or clean manifest (to use for caching)
+        manifest = use_cache and self.cached_manifest or PackageManifest()
+        cached_sources = manifest.sources or {}
+        cached_contract_types = manifest.contract_types or {}
+        sources = {s for s in self.sources if s in file_paths} if file_paths else self.sources
+
+        # Filter out deleted sources
+        deleted_source_ids = cached_sources.keys() - set(
+            map(str, [get_relative_path(s, self.contracts_folder) for s in sources])
+        )
+        contract_types = {
+            name: contract_type
+            for name, contract_type in cached_contract_types.items()
+            if contract_type.source_id not in deleted_source_ids
+        }
+
+        def file_needs_compiling(source: Path) -> bool:
+            path = str(get_relative_path(source, self.contracts_folder))
+
+            if path not in cached_sources:
+                return True  # New file added
+
+            cached = cached_sources[path]
+            cached.compute_checksum(algorithm="md5")
```

Comment:
Hmm, don't really like this (but I know `ethpm-types` would have to be updated)

Feel like it shoudl be `cached.checksum = compute_checksum(cached, algorithm="md5")`

Reply:
We can do:

```python
cached.checksum = Checksum(  # type: ignore
    hash=compute_checksum(cached.content.encode("utf8"), algorithm="md5"),
    algorithm="md5",
)
```
without waiting for an update

Reply:
However, `cachced.content` is optional and `mypy` does not like that

Reply:
Can work on this later, just seems like a method that should be more pure. The in-place mutation always skeeves me out